mindmeld/auto_annotator.py

```diff
@@ -835,5 +901,320 @@ def valid_entity_check(self, entity):
+            )
+        )
+        spans = [
+            Span(candidate["start"], candidate["end"] - 1)
+            for candidate in filtered_candidates
+        ]
+        final_spans = (
+            NoTranslationDucklingAnnotator._get_largest_non_overlapping_candidates(
+                spans
+            )
+        )
+        final_candidates = []
+        for span in final_spans:
+            for candidate in filtered_candidates:
+                if span == Span(candidate["start"], candidate["end"] - 1):
+                    final_candidates.append(candidate)
+                    break
+        if entity_types:
+            final_candidates = [
+                e for e in final_candidates if e["entity_type"] in entity_types
+            ]
+        return [
+            duckling_item_to_entity_dict(candidate) for candidate in final_candidates
+        ]
+
+    @property
+    def supported_entity_types(self):  # pylint: disable=W0236
+        """
+        Returns:
+            supported_entity_types (list): List of supported entity types.
+        """
+        return DUCKLING_TO_SYS_ENTITY_MAPPINGS[self.language]
+
+    @staticmethod
+    def _get_largest_non_overlapping_candidates(spans):
+        """Finds the set of the largest non-overlapping candidates.
+
+        Args:
+            spans (list): List of tuples representing candidate spans (start_index, end_index + 1).
+        Returns:
+            selected_spans (list): List of the largest non-overlapping spans.
+        """
+        spans.sort(reverse=True)
+        selected_spans = []
+        for span in spans:
+            has_overlaps = [
+                span.has_overlap(selected_span) for selected_span in selected_spans
+            ]
+            if not any(has_overlaps):
+                selected_spans.append(span)
+        return selected_spans
+
+    @staticmethod
+    def _filter_out_bad_duckling_candidates(candidates):
+        """Pipeline function to filter initial list of duckling candidates using heuristics.
+
+        Args:
+            candidates (list): List of duckling candidates
+        Returns:
+            filtered_candidates (list): List of filtered duckling candidates.
+        """
+        filtered_candidates = (
+            NoTranslationDucklingAnnotator._remove_unresolved_sys_amount_of_money(
+                candidates
+            )
+        )
+        return filtered_candidates
+
+    @staticmethod
+    def _remove_unresolved_sys_amount_of_money(candidates):
+        """Do not label candidate entities that are sys_amount-of-money but
+        do not have an "unknown" unit type.
+        """
+        return [
+            candidate
+            for candidate in candidates
+            if not (
+                candidate["dim"] == "amount-of-money"
+                and candidate["value"].get("unit") == "unknown"
+            )
+        ]
+
+
+class TranslationDucklingAnnotator(Annotator):
+    """Custom Annotator class used to generate annotations."""
```

Comment:
More detailed docs here on what TranslationDucklingAnnotator does would be helpful